hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledTaskDescriptor.java

```diff
@@ -90,6 +117,79 @@ void setScheduledFuture(ScheduledFuture<?> future) {
         this.scheduledFuture = future;
     }
 
+    void setResult(Object result) {
+        this.result.set(result);
+    }
+
+    Object getResult()
+            throws ExecutionException, InterruptedException {
+
+        return unpeelFutureOrRoot().get();
+    }
+
+    boolean cancel(boolean mayInterrupt)
+            throws ExecutionException, InterruptedException {
+
+        if (!cancelled.compareAndSet(false, true) || scheduledFuture == null) {
+            return false;
+        }
+
+        // Unpeel Executor's Future (in case of Callable, see. DelegatingCallableTaskDecorator)
+        Future delegatedFuture = null;
+        if (SINGLE_RUN.equals(definition.getType())) {
+            delegatedFuture = (Future) scheduledFuture.get();
+        }
+
+        // Try cancel the scheduled to skip further runs
+        boolean cancelled = scheduledFuture.cancel(mayInterrupt);
+        if (delegatedFuture != null) {
+            // Try cancel the delegated Future (Scheduled Callable) to cancel current run.
+            cancelled = delegatedFuture.cancel(mayInterrupt);
+        }
+
+        return cancelled;
+    }
+
+    long getDelay(TimeUnit unit) {
+        boolean isDone = result.get() != null;
+        if (cancelled.get() || isDone) {
+            return 0;
+        }
+
+        return scheduledFuture.getDelay(unit);
+    }
+
+    boolean isCancelled()
+            throws ExecutionException, InterruptedException {
+
+        boolean cancelled = this.cancelled.get();
+        return cancelled || (scheduledFuture != null && unpeelFutureOrRoot().isCancelled());
+    }
+
+    boolean isDone()
+            throws ExecutionException, InterruptedException {
+
+        boolean finished = this.result.get() != null;
+        boolean cancelled = this.cancelled.get();
+        return finished || cancelled || (scheduledFuture != null && unpeelFutureOrRoot().isDone());
+    }
+
+    boolean shouldSchedule() {
+        // Backup tasks (stashed) that aren't Callable's that finished and successfully synced, neither cancelled on another node.
+        return scheduledFuture == null && this.result.get() == null && !this.cancelled.get();
```

Comment:
What if the user scheduled a callable that completed and returned `null`? Does that mean that the callable get rescheduled? Or when the task threw an exception? Related, check other places where `result.get()` is compared to `null`.